md/data/importer.py

```diff
@@ -1,56 +1,173 @@
 import logging
 import os
+import re
+
 import pandas as pd
 
 from django.conf import settings
-from django.db import connections
 
-from tsdata.util import call, line_count, download_and_unzip_data
+from tsdata.util import (call, download_and_unzip_data, get_csv_path,
+     get_datafile_path, line_count)
 
 
 logger = logging.getLogger(__name__)
-cursor = connections['traffic_stops_nc'].cursor()
+
+TIME_OF_STOP_re = re.compile(r'(\d?\d):(\d\d)( [AP]M)?$')
+DEFAULT_TIME_OF_STOP = '00:00'
+
+GENDER_MALE_re = re.compile(r'^(M|male|MALE|m +)$')
+GENDER_FEMALE_re = re.compile(r'^(F|female|w|F +)$')
+
+SEIZED_CONTRABAND_re = re.compile(r'^(Contraband.*|paraphernalia.*|Both)$')
+
+ETHNICITY_WHITE_re = re.compile(r'^(WHITE|W|W.)$')
+ETHNICITY_BLACK_re = re.compile(r'^(BLACK|BLK)$')
+ETHNICITY_TO_CODE = {
+    'HISPANIC': 'H',
+    'ASIAN': 'A',
+    'NATIVE AMERICAN': 'I',
+}
+
+STOP_REASON_re = re.compile(r'^ *(\d+) *- *(\d+) *\(.*\) *$')
+
+MD_COLUMNS_TO_DROP = (
+    'WHATSEARCHED', 'STOPOUTCOME', 'CRIME_CHARGED',
+    'REGISTRATION_STATE', 'RESIDENCE_STATE', 'MD_COUNTY',
+)
+
+
+def fix_ETHNICITY(s):
+    if ETHNICITY_WHITE_re.match(s):
+        return 'W'
+    elif ETHNICITY_BLACK_re.match(s):
+        return 'B'
+    else:
+        return ETHNICITY_TO_CODE.get(s, 'U')
+
+
+def fix_GENDER(s):
+    if GENDER_MALE_re.match(s):
+        return 'M'
+    elif GENDER_FEMALE_re.match(s):
+        return 'F'
+    else:
+        return 'U'
+
+
+def fix_SEIZED(s):
+    if SEIZED_CONTRABAND_re.match(s):
+        return 'Y'
+    else:
+        return 'N'
+
+
+def fix_STOP_REASON(s):
+    m = STOP_REASON_re.match(s)
+    if m:
+        return m.group(1) + '-' + m.group(2)
+    else:
+        return s
+
+
+def fix_TIME_OF_STOP(s):
+    s = s.strip()
+    m = TIME_OF_STOP_re.match(s)
+    if not m:
+        return DEFAULT_TIME_OF_STOP
+    hour = int(m.group(1))
+    minute = int(m.group(2))
+    if not 0 <= hour < 24:
+        return DEFAULT_TIME_OF_STOP
+    if not 0 <= minute < 60:
+        return DEFAULT_TIME_OF_STOP
+
+    # XXX Check converted CSV to see if we need to fix up the handling
+    #     of 'PM', which can be appended to times before or after 12:00 p.m.
+    # IOW, does the right thing happen for both '8:53 PM' and '20:53 PM'?
```

Comment:
I wonder if this could be unit tested too. These apply functions are perfect to test a few cases of bad data.